File: agents/src/stability.py
def aggregate_prompt_scores(prompt_scores: list[dict], runs: list[dict]) -> list[dict]:
    by_run_query: dict[str, dict[str, list[dict]]] = {}
    for ps in prompt_scores:
        rid = ps["run_id"]
        q = ps["query"]
        by_run_query.setdefault(rid, {}).setdefault(q, []).append(ps)

    result = []
    for run in runs:
        rid = run["id"]
        queries = {}
        for query, scores in by_run_query.get(rid, {}).items():
            n = len(scores)
            avg_rate = sum(s["mention_rate"] for s in scores) / n if n > 0 else 0

            total_weight = sum(s["mention_rate"] for s in scores)
            if total_weight > 0:
                avg_level = round(
                    sum(s["avg_mention_level"] * s["mention_rate"] for s in scores) / total_weight,
                    10,
                )
            else:
                avg_level = 0

            queries[query] = {
                "mention_rate": avg_rate,
                "avg_mention_level": avg_level,
            }

        result.append({
            "run_id": rid,
            "ran_at": run["ran_at"],
            "queries": queries,
        })

    return result


def _classify(rates: list[float], levels: list[float]) -> str:
    n = len(rates)

    if all(r == 0 for r in rates):
        return "absent"

    if n >= 3 and all(r >= 0.7 for r in rates):
        level_range = max(levels) - min(levels)
        if level_range <= 0.5:
            return "locked_in"

    if n >= 3:
        rate_diffs = [rates[i + 1] - rates[i] for i in range(n - 1)]
        signs = [1 if d > 0 else (-1 if d < 0 else 0) for d in rate_diffs]
        nonzero_signs = [s for s in signs if s != 0]
        if len(set(nonzero_signs)) > 1:
            return "volatile"

    if n >= 2:
        rate_delta = rates[-1] - rates[0]
        level_delta = levels[-1] - levels[0]

        rate_gaining = rate_delta >= 0.1
        rate_declining = rate_delta <= -0.1
        level_gaining = level_delta >= 0.5
        level_declining = level_delta <= -0.5

        if rate_gaining or (not rate_declining and level_gaining):
            return "gaining"

        if rate_declining or (not rate_gaining and level_declining):
            return "declining"

    return "volatile"


def compute_prompt_stability(runs_data: list[dict]) -> list[dict]:
    if not runs_data:
        return []

    all_queries: set[str] = set()
    for run in runs_data:
        all_queries.update(run["queries"].keys())

    result = []
    for query in sorted(all_queries):
        rates = []
        levels = []
        trend = []

        for run in runs_data:
            q_data = run["queries"].get(query, {"mention_rate": 0, "avg_mention_level": 0})
            rates.append(q_data["mention_rate"])
            levels.append(q_data["avg_mention_level"])
            trend.append({
                "run_id": run["run_id"],
                "ran_at": run["ran_at"],
                "mention_rate": q_data["mention_rate"],
                "avg_mention_level": q_data["avg_mention_level"],
            })

        stability_class = _classify(rates, levels)

        result.append({
            "query": query,
            "stability_class": stability_class,
            "current_mention_rate": rates[-1],
            "current_avg_level": levels[-1],
            "trend": trend,
        })

    return result

File: dashboard/lib/stability.ts
export type StabilityClass = "locked_in" | "gaining" | "declining" | "volatile" | "absent";

interface QueryScores {
  query_id: string | null;
  query: string;
  mention_rate: number;
  avg_mention_level: number;
}

interface RunData {
  run_id: string;
  ran_at: string;
  queries: Record<string, QueryScores>;
}

export interface PromptStabilityResult {
  query_id: string | null;
  query: string;
  stability_class: StabilityClass;
  current_mention_rate: number;
  current_avg_level: number;
  trend: {
    run_id: string;
    ran_at: string;
    mention_rate: number;
    avg_mention_level: number;
  }[];
}

function classify(rates: number[], levels: number[]): StabilityClass {
  const n = rates.length;

  if (rates.every((r) => r === 0)) {
    return "absent";
  }

  if (n >= 3 && rates.every((r) => r >= 0.7)) {
    const levelRange = Math.max(...levels) - Math.min(...levels);
    if (levelRange <= 0.5) {
      return "locked_in";
    }
  }

  if (n >= 3) {
    const rateDiffs = [];
    for (let i = 0; i < n - 1; i++) {
      rateDiffs.push(rates[i + 1] - rates[i]);
    }
    const signs = rateDiffs.map((d) => (d > 0 ? 1 : d < 0 ? -1 : 0));
    const nonzeroSigns = signs.filter((s) => s !== 0);
    const uniqueSigns = new Set(nonzeroSigns);
    if (uniqueSigns.size > 1) {
      return "volatile";
    }
  }

  if (n >= 2) {
    const rateDelta = rates[n - 1] - rates[0];
    const levelDelta = levels[n - 1] - levels[0];

    const rateGaining = rateDelta >= 0.1;
    const rateDeclining = rateDelta <= -0.1;
    const levelGaining = levelDelta >= 0.5;
    const levelDeclining = levelDelta <= -0.5;

    if (rateGaining || (!rateDeclining && levelGaining)) {
      return "gaining";
    }

    if (rateDeclining || (!rateGaining && levelDeclining)) {
      return "declining";
    }
  }

  return "volatile";
}

export function aggregatePromptScores(
  promptScores: { run_id: string; query_id?: string | null; query: string; llm: string; mention_rate: number; avg_mention_level: number }[],
  runs: { id: string; ran_at: string }[]
): RunData[] {
  const byRunQuery: Record<string, Record<string, typeof promptScores>> = {};

  for (const ps of promptScores) {
    const key = ps.query_id || ps.query;
    if (!byRunQuery[ps.run_id]) byRunQuery[ps.run_id] = {};
    if (!byRunQuery[ps.run_id][key]) byRunQuery[ps.run_id][key] = [];
    byRunQuery[ps.run_id][key].push(ps);
  }

  return runs.map((run) => {
    const queries: Record<string, QueryScores> = {};
    const runScores = byRunQuery[run.id] || {};

    for (const [query, scores] of Object.entries(runScores)) {
      const n = scores.length;
      const avgRate = scores.reduce((s, x) => s + x.mention_rate, 0) / n;
      const label = scores[scores.length - 1]?.query || query;
      const queryId = scores.find((x) => x.query_id)?.query_id || null;

      const totalWeight = scores.reduce((s, x) => s + x.mention_rate, 0);
      const avgLevel =
        totalWeight > 0
          ? scores.reduce((s, x) => s + x.avg_mention_level * x.mention_rate, 0) / totalWeight
          : 0;

      queries[query] = {
        query_id: queryId,
        query: label,
        mention_rate: avgRate,
        avg_mention_level: avgLevel,
      };
    }

    return { run_id: run.id, ran_at: run.ran_at, queries };
  });
}

export function computePromptStability(runsData: RunData[]): PromptStabilityResult[] {
  if (runsData.length === 0) return [];

  const allQueries = new Set<string>();
  for (const run of runsData) {
    for (const q of Object.keys(run.queries)) {
      allQueries.add(q);
    }
  }

  const results: PromptStabilityResult[] = [];

  for (const query of [...allQueries].sort()) {
    const rates: number[] = [];
    const levels: number[] = [];
    const trend: PromptStabilityResult["trend"] = [];
    let label = query;
    let queryId: string | null = null;

    for (const run of runsData) {
      const qData = run.queries[query];
      if (qData) {
        label = qData.query;
        queryId = qData.query_id;
      }
      const metrics = qData || { mention_rate: 0, avg_mention_level: 0 };
      rates.push(metrics.mention_rate);
      levels.push(metrics.avg_mention_level);
      trend.push({
        run_id: run.run_id,
        ran_at: run.ran_at,
        mention_rate: metrics.mention_rate,
        avg_mention_level: metrics.avg_mention_level,
      });
    }

    results.push({
      query_id: queryId,
      query: label,
      stability_class: classify(rates, levels),
      current_mention_rate: rates[rates.length - 1],
      current_avg_level: levels[levels.length - 1],
      trend,
    });
  }

  return results;
}
